### PR TITLE
Disable autocast in GroupNormalization for mixed precision stability

### LINKED ISSUE #22586: [Bug] Mixed Precision (float16) numerical instability in GroupNormalization with small epsilon
## Description

While using keras.mixed_precision.set_global_policy("mixed_float16"), the GroupNormalization layer produces NaN outputs when the epsilon parameter is set to a value smaller than the precision limit of float16 (e.g., $1e-12$). Since the smallest positive normalized value in FP16 is approximately $6.1e-5$, a very small epsilon effectively acts as zero during variance normalization, leading to numerical collapse.

## Poc

```
import os
os.environ["KERAS_BACKEND"] = "tensorflow"
import keras
import numpy as np

keras.mixed_precision.set_global_policy("mixed_float16")
inputs = keras.Input(shape=(16, 16, 64))
# Epsilon 1e-12 causes NaN in FP16
x = keras.layers.GroupNormalization(groups=8, epsilon=1e-12)(inputs)
model = keras.Model(inputs, x)

extreme_data = np.random.uniform(low=60000, high=70000, size=(1, 16, 16, 64)).astype("float32")
output = model(extreme_data)
print(f"NaN detected: {np.any(np.isnan(output))}")


```

## Observed Result

```
NaN detected: True

```

## Expected Behavior

The layer should either:

* Automatically clip the epsilon to a safe minimum (e.g., 1e-7) when the policy is float16.

* Perform the normalization internal math in float32 regardless of the policy (Upcasting).

## Actual Behavior

The layer produces NaN values immediately when exposed to large input values or small epsilons in mixed_float16 mode.

## Notes
* This issue arises from float16 precision limits rather than user misuse.
* Other normalization layers (e.g., BatchNormalization) internally upcast to float32 to avoid similar issues.

### PR DESCRIPTION
**Pull request by Vedag812**
Fixes #22586

## Contributor Agreement

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

## The problem

Under `mixed_float16`, `GroupNormalization` produces NaN outputs when inputs contain values above the float16 max (~65504). This is because `autocast=True` (the default) causes the layer's `__call__()` to cast inputs to float16 **before** `call()` runs. Values above 65504 overflow to `inf` at that point, and the internal float32 upcast inside `_apply_normalization` can't recover them since they are already `inf`.

This was discussed in #22586 and a previous PR (#22666) attempted a different approach (moving the upcast earlier in `call()`) but was closed after going stale without a signed CLA. The last comment on #22586 from @Lawson-Darrow identified the correct minimal fix.

## The fix

One line: `self.autocast = False` in `GroupNormalization.__init__()`.

This is the exact same approach `LayerNormalization` already uses (see [line 151](https://github.com/keras-team/keras/blob/master/keras/src/layers/normalization/layer_normalization.py#L151)). With autocast disabled, inputs keep their original dtype (float32) when entering `call()`, so the existing float32 upcast in `_apply_normalization` works as intended.

## Why this is the right approach

The Keras maintainer @hertschuh [commented](https://github.com/keras-team/keras/pull/22666#issuecomment-2825373637) that the fix should be to let users pass `autocast=False`. However, `LayerNormalization` already sets this unconditionally because normalization layers are inherently sensitive to precision. `GroupNormalization` should be consistent with that. Users shouldn't have to know about this footgun.

## What changed

| File | Change |
|------|--------|
| `group_normalization.py` | Added `self.autocast = False` in `__init__()` |
| `group_normalization_test.py` | Added regression test with large inputs under `mixed_float16` |

## Test

The new `test_mixed_float16_no_nan` creates a `GroupNormalization` layer with `dtype='mixed_float16'`, feeds it values in the 60000-70000 range (spanning the float16 max), and asserts no NaN in the output. Without the fix, this test fails.

### CODE REVIEW

**Comment by Lawson-Darrow** [comment]
I offered to open this pr in the comment I laid out the fix in. Given the discussion around the issue and the contribution policy I though it best to wait for a maintainer to weigh in first before opening the pr.

**Comment by Vedag812** [comment]
Hey @Lawson-Darrow, apologies for jumping ahead on this. Your analysis in the issue comment was spot on and directly informed this PR. I should have checked more carefully whether you were planning to submit it yourself.

If you'd prefer to open the PR from your end, I'm happy to close this one. Alternatively, if the maintainers are fine with this PR moving forward, I wanted to make sure you get proper credit for identifying the root cause and the correct fix (setting autocast=False to match LayerNormalization).

Either way works for me, let me know what you'd prefer.

**Comment by Lawson-Darrow** [comment]
Thanks, no worries. 

Since you offered, I'll open the PR from my end so it ties back to the issue analysis and gives a maintainer the ability to weigh in first.  

Appreciate you catching it and feel free to review once it's up.

**Comment by Lawson-Darrow** [comment]
Pr is posted.

#23070

**Comment by hertschuh** [comment]
Closing in favor of https://github.com/keras-team/keras/pull/23070
